File: iii_drone_configuration/configuration.py
from __future__ import annotations

from dataclasses import dataclass
from threading import Lock
from typing import Any, Callable
import rclpy

@dataclass
class ConfigurationEntry:
    full_name: str
    parameter_type: rclpy.parameter.Parameter.Type


class Configuration:
    def __init__(
        self,
        native_configuration: Any,
        parameter_getter: Callable[[str], rclpy.parameter.Parameter],
    ):
        self._native_configuration = native_configuration
        self._parameter_getter = parameter_getter
        self._lock = Lock()

    def get_parameter(self, parameter_full_name: str) -> rclpy.parameter.Parameter:
        with self._lock:
            if self._native_configuration.has_parameter(parameter_full_name):
                return self._parameter_getter(parameter_full_name)

        raise RuntimeError(f"Parameter {parameter_full_name} does not exist.")

    def has_parameter(self, parameter_full_name: str) -> bool:
        with self._lock:
            return self._native_configuration.has_parameter(parameter_full_name)

    @property
    def name(self) -> str:
        return self._native_configuration.name

File: src/configuration.cpp
#include <iii_drone_configuration/configuration.hpp>

using namespace iii_drone::configuration;

Configuration::Configuration(
    std::string name,
    std::vector<configuration_entry_t> configuration_entries,
    std::function<rclcpp::Parameter(const std::string &)> parameter_getter
) : name_(std::move(name)),
    parameter_getter_(std::move(parameter_getter)),
    configuration_entries_(std::move(configuration_entries))
{
}

rclcpp::Parameter Configuration::GetParameter(const std::string & parameter_full_name) const
{
    std::shared_lock<std::shared_mutex> lock(mutex_);

    for (const auto & configuration_entry : configuration_entries_) {
        if (configuration_entry.full_name == parameter_full_name) {
            return parameter_getter_(configuration_entry.full_name);
        }
    }

    throw std::runtime_error(
        "Configuration::GetParameter(): Parameter " + parameter_full_name + " does not exist."
    );
}

bool Configuration::HasParameter(const std::string & parameter_full_name) const
{
    std::shared_lock<std::shared_mutex> lock(mutex_);

    for (const auto & configuration_entry : configuration_entries_) {
        if (configuration_entry.full_name == parameter_full_name) {
            return true;
        }
    }

    return false;
}

std::string Configuration::name() const
{
    return name_;
}
